agent/recorder.go

```diff
@@ -120,89 +151,161 @@ func (r *SpanRecorder) loop() error {
 	if err != nil {
-		r.koSend++
 		return 0, err
 	}
 	req.Header.Set("User-Agent", r.userAgent)
 	req.Header.Set("Content-Type", "application/msgpack")
 	req.Header.Set("Content-Encoding", "gzip")
 	req.Header.Set("X-Scope-ApiKey", r.apiKey)
 
-	resp, err := r.client.Do(req)
-	if err != nil {
-		r.koSend++
-		return 0, err
-	}
-	defer resp.Body.Close()
+	var lastError error
+	for i := 0; i <= numOfRetries; i++ {
+		if r.debugMode {
+			if i == 0 {
+				r.logger.Println("sending payload")
+			} else {
+				r.logger.Printf("sending payload [retry %d]", i)
+			}
+		}
 
-	if resp.StatusCode >= 400 {
-		body, err := ioutil.ReadAll(resp.Body)
+		resp, err := r.client.Do(req)
 		if err != nil {
-			return resp.StatusCode, err
+			if v, ok := err.(*url.Error); ok {
+				// Don't retry if the error was due to too many redirects.
+				if redirectsErrorRe.MatchString(v.Error()) {
+					return 0, errors.New(fmt.Sprintf("error: http client returns: %v", err.Error()))
+				}
+
+				// Don't retry if the error was due to an invalid protocol scheme.
+				if schemeErrorRe.MatchString(v.Error()) {
+					return 0, errors.New(fmt.Sprintf("error: http client returns: %v", err.Error()))
+				}
+
+				// Don't retry if the error was due to TLS cert verification failure.
+				if _, ok := v.Err.(x509.UnknownAuthorityError); ok {
+					return 0, errors.New(fmt.Sprintf("error: http client returns: %v", err.Error()))
+				}
+			}
+
+			lastError = err
+			r.logger.Printf("error: client timeout, retrying in %d seconds", retryBackoff/time.Second)
+			time.Sleep(retryBackoff)
+			atomic.AddInt64(&r.stats.sendSpansRetries, 1)
+			continue
+		}
+
+		var (
+			bodyData []byte
+			status   string
+		)
+		statusCode = resp.StatusCode
+		status = resp.Status
+		if resp.Body != nil && resp.Body != http.NoBody {
+			body, err := ioutil.ReadAll(resp.Body)
+			if err == nil {
+				bodyData = body
+			}
+		}
+		if err := resp.Body.Close(); err != nil { // We can't defer inside a for loop
+			r.logger.Printf("error: closing the response body. %s", err.Error())
+		}
+
+		if statusCode == 0 || statusCode >= 400 {
+			lastError = errors.New(fmt.Sprintf("error from API [status: %s]: %s", status, string(bodyData)))
+		}
+
+		// Check the response code. We retry on 500-range responses to allow
+		// the server time to recover, as 500's are typically not permanent
+		// errors and may relate to outages on the server side. This will catch
+		// invalid response codes as well, like 0 and 999.
+		if statusCode == 0 || (statusCode >= 500 && statusCode != 501) {
+			r.logger.Printf("error: status code: %d, retrying in %d seconds", statusCode, retryBackoff/time.Second)
+			time.Sleep(retryBackoff)
+			atomic.AddInt64(&r.stats.sendSpansRetries, 1)
+			continue
 		}
-		return resp.StatusCode, errors.New(fmt.Sprintf("error from API (%s): %s", resp.Status, body))
+
+		if statusCode < 400 {
+			return statusCode, nil
+		}
+		return statusCode, lastError
```

Comment:
Remove this, right?

Reply:
Why? if there is no error or no retries, then I have to return. Maybe the option is not to remove it, but changed it with a break, so we exit the for-loop

Reply:
I've changed with a break